### PR TITLE
Support custom implicit Function Arguments like PosInfos via abstract @:fromNothing macro

### PR DESCRIPTION
**Pull request by frabbit**

### CODE REVIEW

File: src/typing/typer.ml [hunk 1]
```diff
@@ -1571,13 +1585,20 @@ let type_bind ctx (e : texpr) (args,ret) params p =
...
+				ordered_args @ [t]
+			| _ ->
+				if is_pos_infos t then
 					let infos = mk_infos ctx p [] in
 					ordered_args @ [type_expr ctx infos (WithType t)]
 				else if ctx.com.config.pf_pad_nulls then
 					(ordered_args @ [(mk (TConst TNull) t_dynamic p)])
 				else
-					ordered_args
+					ordered_args @ [(mk (TConst TNull) t_dynamic p)]
```
Comment (frabbit) [inline src/typing/typer.ml]: @ncannasse does this change, actually ignoring pf_pad_null for bind, does have any effects on real code? The unit tests still they pass.

Comment (ncannasse) [reply src/typing/typer.ml]: well it's not ignoring pf_pad_null but is always pushing nulls whereas you're padding or not. I'm not sure I like it, but I guess we can remove extra nulls at a later stage maybe?

Comment (frabbit) [reply src/typing/typer.ml]: yes, true. I actually wonder if and what implications this actually have.

Comment (frabbit) [reply src/typing/typer.ml]: or to be more precise, which negative implications this have.

Comment (frabbit) [reply src/typing/typer.ml]: and just to let you know, there are already cases where this leads to bugs: http://try-haxe.mrcdk.com/#12d04.

**Comment by Simn** [comment]
You will have to rebase this one.

**Comment by frabbit** [comment]
It's up to date again

**Comment by frabbit** [comment]
@ncannasse @Simn @nadako any objections or remarks?

**Comment by nadako** [comment]
I'm yet to check this out and see what can I do with it :)

**Comment by ncannasse** [comment]
I think fromNothing is badly named.  I suggest using `defaultValue()` instead. Also, does it works if the method is not a macro as well ?

**Comment by Simn** [comment]
> I think fromNothing is badly named. I suggest using defaultValue() instead. Also, does it works if the method is not a macro as well ?

I agree that it's badly named. But `defaultValue` is not much better. We should make it clear that this is only for function arguments so people don't assume that it's run on something like `var x:Implicit<Blub>;`.

**Comment by frabbit** [comment]
@ncannasse 
> Also, does it works if the method is not a macro as well ?

Yes, it does atm, but its not very useful in my opinion because you don't know the expected type.

**Comment by back2dos** [comment]
What about fields of object literals? Seems like a relatively good use case (will give you prop drilling for example). Would there be anything wrong with also using this for initializing variables?

**Comment by frabbit** [comment]
@back2dos would love to see an example, because i fail to see how that could be done without having an abstract with a concrete underlying type.

**Comment by RealyUniqueName** [comment]
Let's revisit this for 4.1

**Comment by Simn** [comment]
This branch would need an update. @RealyUniqueName Could you look into this?

**Comment by RealyUniqueName** [comment]
Sure

**Comment by Simn** [comment]
This PR is unfortunately unmergeable at this point. I'll link it to #4583 mostly because there are some good tests here, so if we add this feature for Haxe 5 we can add those in some way.

**Pull request closed without merging**